protocol/messenger.go

```diff
@@ -1778,88 +1785,161 @@ func (m *Messenger) handlePushNotificationClientRegistrations(c chan struct{}) {
+	exists, err := m.communitiesManager.CommunitySettingsExist(org.ID())
+	if err != nil {
+		return err
+	}
 
-		// In case we do have settings, but the history archive support is disabled
-		// for this community, we enable it, as this should be the default for all
-		// non-admin communities
-		communitySettings, err := m.communitiesManager.GetCommunitySettingsByID(org.ID())
-		if err != nil {
-			logger.Warn("failed to fetch community settings", zap.Error(err))
-			continue
+	if !exists {
+		communitySettings := communities.CommunitySettings{
+			CommunityID:                  org.IDString(),
+			HistoryArchiveSupportEnabled: true,
 		}
+		return m.communitiesManager.SaveCommunitySettings(communitySettings)
+	}
 
-		if !org.IsControlNode() && !communitySettings.HistoryArchiveSupportEnabled {
-			communitySettings.HistoryArchiveSupportEnabled = true
-			err = m.communitiesManager.UpdateCommunitySettings(*communitySettings)
-			if err != nil {
-				logger.Warn("failed to update community settings", zap.Error(err))
-			}
-		}
+	// In case we do have settings, but the history archive support is disabled
+	// for this community, we enable it, as this should be the default for all
+	// non-admin communities
+	communitySettings, err := m.communitiesManager.GetCommunitySettingsByID(org.ID())
+	if err != nil {
+		return err
 	}
 
+	if !org.IsControlNode() && !communitySettings.HistoryArchiveSupportEnabled {
+		communitySettings.HistoryArchiveSupportEnabled = true
+		return m.communitiesManager.UpdateCommunitySettings(*communitySettings)
+	}
+
+	return nil
+}
+
+func (m *Messenger) processSpectatedCommunities(wg *sync.WaitGroup, filtersCh chan<- []transport.FiltersToInitialize, errCh chan<- error) {
+	defer gocommon.LogOnPanic()
+	defer wg.Done()
+
 	spectatedCommunities, err := m.communitiesManager.Spectated()
 	if err != nil {
-		return err
+		errCh <- err
+		return
 	}
+
+	var filtersToInit []transport.FiltersToInitialize
 	for _, org := range spectatedCommunities {
 		filtersToInit = append(filtersToInit, m.DefaultFilters(org)...)
 	}
+	filtersCh <- filtersToInit
+}
+
+func (m *Messenger) processChats(wg *sync.WaitGroup, filtersCh chan<- []transport.FiltersToInitialize, publicKeysCh chan<- []*ecdsa.PublicKey, errCh chan<- error) {
+	defer gocommon.LogOnPanic()
+	defer wg.Done()
 
 	// Get chat IDs and public keys from the existing chats.
 	// TODO: Get only active chats by the query.
 	chats, err := m.persistence.Chats()
 	if err != nil {
-		return err
+		errCh <- err
+		return
+	}
+
+	validChats, communityInfo := m.validateAndProcessChats(chats)
+	filters, publicKeys, err := m.processValidChats(validChats, communityInfo)
+	if err != nil {
+		errCh <- err
+		return
 	}
 
+	filtersCh <- filters
+	publicKeysCh <- publicKeys
+
+	if err := m.processDeprecatedChats(); err != nil {
+		errCh <- err
+	}
+}
+
+func (m *Messenger) validateAndProcessChats(chats []*Chat) ([]*Chat, map[string]*communities.Community) {
```

Comment:
This function doesn't seem to "process" chats.
```suggestion
func (m *Messenger) validateChats(chats []*Chat) ([]*Chat, map[string]*communities.Community) {
```

Reply:
updated it to `validateAndInitChatsFirstMessageTimestamp`